Objects/listobject.c

```diff
@@ -180,6 +180,22 @@ PyList_New(Py_ssize_t size)
     return (PyObject *) op;
 }
 
+static PyObject *
+list_new_prealloc(Py_ssize_t size)
+{
+    PyListObject *op = (PyListObject *) PyList_New(0);
+    if (size == 0 || op == NULL) {
+        return (PyObject *) op;
+    }
+    op->ob_item = PyMem_New(PyObject *, size);
+    if (op->ob_item == NULL) {
+        Py_DECREF(op);
+        return PyErr_NoMemory();
+    }
```

Comment:
From this point, it is possible that a GC collection will access the list. It would be safer to *not* call _PyObject_GC_TRACK(op) before the list is fully initialized. I suggest to modify list_new_prealloc() to avoid _PyObject_GC_TRACK(op), but call in the caller.

Maybe you can rewrite the code so PyList_New() calls list_new_prealloc(), and not the opposite?

Reply:
Could you elaborate on what do you mean exactly by not "fully initialized" here?

I think I could rewrite the whole patch using existing `list_resize()` function (I didn't it initially because unlike `PyList_New()` it preallocates more space than needed), since its usage doesn't raise such questions.

Reply:
list_dealloc() will crash if it's called on the list just after list_new_prealloc(), since items are uninitialized, whereas list_dealloc() calls Py_XDECREF() on each item. But the main issue is list_traverse(). If the garbage collection is triggered, list_traverse() will dereference uninitialized object pointers and so crash.

Reply:
list_traverse doesn’t touch preallocated area.
Unless Py_SIZE(list) is updated before filling objects, there are no dangers, I suppose.